src/include/sof/lib/mm_heap.h

```diff
@@ -0,0 +1,95 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2016 Intel Corporation. All rights reserved.
+ *
+ * Author: Liam Girdwood <liam.r.girdwood@linux.intel.com>
+ *         Keyon Jie <yang.jie@linux.intel.com>
+ */
+
+#ifndef __SOF_LIB_MM_HEAP_H__
+#define __SOF_LIB_MM_HEAP_H__
+
+#include <sof/common.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/cache.h>
+#include <sof/lib/memory.h>
+#include <sof/sof.h>
+#include <sof/spinlock.h>
+#include <config.h>
+#include <stddef.h>
+#include <stdint.h>
+
+struct dma_copy;
+struct dma_sg_config;
+
+struct mm_info {
+	uint32_t used;
+	uint32_t free;
+};
+
+struct block_hdr {
+	uint16_t size;		/* size in blocks for continuous allocation */
+	uint16_t used;		/* usage flags for page */
+	void *unaligned_ptr;	/* align ptr */
+} __packed;
+
+struct block_map {
+	uint16_t block_size;	/* size of block in bytes */
+	uint16_t count;		/* number of blocks in map */
+	uint16_t free_count;	/* number of free blocks */
+	uint16_t first_free;	/* index of first free block */
+	struct block_hdr *block;	/* base block header */
+	uint32_t base;		/* base address of space */
+};
```

Comment:
Probably worth packing this as it was missing from initial implementation.

Reply:
Won't packing cause performance issues? Memory allocation is pretty hot code, even if not directly during playback. And this isn't an IPC structure but an internal one so I'd say packing isn't that good of an idea.

Reply:
I think we are OK in this context since xtensa has 16bit loads and saves and we would be aligned on 16bit for 16bit data. i.e. we are not doing any shifts or unaligned loads.

Reply:
I'm not sure what is benefit for packing here. Saving few bytes for possible performance issues with embedded compilers that do not have quality comparable to x86 gcc compilers and may not be that smart about optimizations.
I'd go for `packed` only for ABI structs.

Reply:
@jajanusz I agree with you when it comes to 8 bit, but 16bit loads have very wide coverage on most RISC architectures. This structure is also repeated often (in arrays) and may be repeated more if we split heaps over 64kB banks. Also 16 bytes vs 24 bytes per element is an improvement for cache too.